### PR TITLE
Augmented Lagrangian Learning Method

### PR DESCRIPTION
**Pull request by andrewrosemberg**
Adds Augmented Lagrangian Primal-Dual Learning Method

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -3,11 +3,29 @@ uuid = "f31bfc7b-7b5d-4cc3-b76b-1af281ce159d"
...
 version = "1.0.0-DEV"
 
+[deps]
+BatchNLPKernels = "7145f916-0e30-4c9d-93a2-b32b6056125d"
+CUDA = "052768ef-5323-5732-b1bb-66c8b64840ba"
+ExaModels = "1037b233-b668-4ce9-9b63-f9f681f55dd2"
+Lux = "b2108857-7c20-44ae-9111-449ecde12c47"
+LuxCUDA = "d0bbae9a-e099-4d5b-a835-1c6931763bda"
+Optimisers = "3bd65402-5787-11e9-1adc-39752487f4e2"
+Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
```
Comment (klamike) [inline Project.toml]: Do we need all of these? In particular CUDA, LuxCUDA, ExaModels?

File: test/runtests.jl [hunk 2]
```diff
@@ -1,6 +1,169 @@
...
+        L2OALM_train!(
+            bm_train,
+            num_equal,
+            primal_model,
+            dual_model,
+            train_state_primal,
+            train_state_dual,
+            data,
+            stopping_criteria = [validation_testset],
+        )
```
Comment (klamike) [inline test/runtests.jl]: Why not put a test like https://github.com/klamike/BatchNLPKernels.jl/blob/b29456ba92dd15748e4c677f0ca3db59a1e8caac/test/test_penalty.jl#L89?

File: src/L2OALM.jl [hunk 3]
```diff
@@ -1,5 +1,426 @@
...
+
+Returns a function that computes the MSE loss for the (dual-)model predicting lagrangian dual variables
+from constraint evaluations `gh` and the current dual predictions `dual_hat_k`.
+Target is calculated using the augmented lagrangian method. 
+Target dual variables are clipped from zero to `max_dual`.
+
+Keywords:
+    - `max_dual`: Maximum value for the target dual variables.
+"""
+function LagrangianDualLoss(num_equal::Int; max_dual = 1e6)
```
Comment (klamike) [inline src/L2OALM.jl]: Just a note, we should probably eventually have a somewhat standard interface for L2OMethods and their hyperparameters, i.e.

```julia
struct ALMMethod <: AbstractL2OMethod
  bm::BatchModel
  max_dual::Float64
  ρ_init::Float64
end
```

or

```julia
struct ALMMethod <: AbstractL2OMethod
  bm::BatchModel
  hyperparameters::Dict{Symbol,Any}
end
```

Ideally that also would help to clean up stuff like https://github.com/LearningToOptimize/L2OALM.jl/blob/0797bb71d848a11b29ba61fa4083792196c4b2b0/src/L2OALM.jl#L196-L197

File: test/runtests.jl [hunk 4]
```diff
@@ -1,6 +1,169 @@
...
+        T = Float64,
+    )
+        model, nbus, ngen, blines =
+            create_parametric_ac_power_model(filename; backend = backend, T = T)
+        bm_train = BNK.BatchModel(model, batch_size, config = BNK.BatchModelConfig(:full))
+        bm_test = BNK.BatchModel(model, dataset_size, config = BNK.BatchModelConfig(:full))
+
+        nvar = model.meta.nvar
+        ncon = model.meta.ncon
+        nθ = length(model.θ)
```
Comment (klamike) [inline test/runtests.jl:68]: This is such a common thing, BNK should probably have a field for `nθ`, and expose a frontend like `num_parameters`, `num_variables`, `num_constraints`.

File: test/runtests.jl [hunk 5]
```diff
@@ -1,6 +1,169 @@
...
+            objs_test = BNK.objective!(bm_test, X̂_test, Θ_test)
+            Vc_test, Vb_test = BNK.all_violations!(bm_test, X̂_test, Θ_test)
+            gh_test = BNK.constraints!(bm_test, X̂_test, Θ_test)
+            dual_hat, _ =
+                dual_model(Θ_test, train_state_dual.parameters, train_state_dual.states)
+            # Separate bound and equality constraints
+            gh_bound = gh_test[1:end-num_equal, :]
+            gh_equal = gh_test[end-num_equal+1:end, :]
+            dual_hat_bound = dual_hat[1:end-num_equal, :]
+            dual_hat_equal = dual_hat[end-num_equal+1:end, :]
```
Comment (klamike) [inline test/runtests.jl]: This is another obvious thing BNK should have -- functions that help you deal with indices

File: test/runtests.jl [hunk 6]
```diff
@@ -1,6 +1,169 @@
...
+            dual_target = vcat(
+                min.(max.(dual_hat_bound + ρ .* gh_bound, 0), max_dual),
+                min.(dual_hat_equal + ρ .* gh_equal, max_dual),
+            )
+
+            dual_loss = mean((dual_hat .- dual_target) .^ 2)
+
+            @info "Validation Testset: Iteration $iter" mean(objs_test) mean(Vc_test) mean(
+                Vb_test,
+            ) dual_loss
```
Comment (klamike) [inline test/runtests.jl]:
```suggestion
            @info "Validation Testset: Iteration $iter" mean(objs_test) mean(Vc_test) mean(Vb_test) dual_loss
```

File: test/runtests.jl [hunk 7]
```diff
@@ -1,6 +1,169 @@
...
+        # lagrangian_prev = sum(y) + 1000 * sum(Vc) + 1000 * sum(Vb)
+
+        train_state_primal =
+            Training.TrainState(primal_model, ps_primal, st_primal, Optimisers.Adam(1e-5))
+        train_state_dual =
+            Training.TrainState(dual_model, ps_dual, st_dual, Optimisers.Adam(1e-5))
+
+        data = DataLoader((Θ_train); batchsize = batch_size, shuffle = true) .|> dev_gpu
+
+        function validation_testset(
```
Comment (klamike) [inline test/runtests.jl]: Maybe the validation function should go in the repo? Also, what about a labeled version for validation? (obviously we still need to write the ExaModels datagen code for it to be useful)

File: src/L2OALM.jl [hunk 8]
```diff
@@ -1,5 +1,426 @@
...
+        X̂, st_new = model(Θ, ps, st)
+
+        # Calculate violations and objectives
+        objs = BNK.objective!(bm, X̂, Θ)
+        # gh = constraints!(bm, X̂, Θ)
+        Vc, Vb = BNK.all_violations!(bm, X̂, Θ)
+        V = vcat(Vb, Vc)
+        total_loss = (
+            sum(abs.(dual_hat .* V)) / num_s + ρ / 2 * sum((V) .^ 2) / num_s + mean(objs)
+        )
```
Comment (klamike) [inline src/L2OALM.jl:202]: should make this a function

File: src/L2OALM.jl [hunk 9]
```diff
@@ -1,5 +1,426 @@
...
+        dual_hat_bound = dual_hat_k[1:end-num_equal, :]
+        dual_hat_equal = dual_hat_k[end-num_equal+1:end, :]
+
+        # Target for dual variables
+        dual_target = vcat(
+            min.(max.(dual_hat_bound + ρ .* gh_bound, 0), max_dual),
+            min.(dual_hat_equal + ρ .* gh_equal, max_dual),
+        )
+
+        loss = mean((dual_hat .- dual_target) .^ 2)
```
Comment (klamike) [inline src/L2OALM.jl:171]: should make this a function

File: src/L2OALM.jl (hunk 8 above)
Comment (klamike) [reply src/L2OALM.jl:202]: also, is `num_s` correct here?

File: src/L2OALM.jl [hunk 10]
```diff
@@ -1,5 +1,426 @@
...
+        total_loss = (
+            sum(abs.(dual_hat .* V)) / num_s + ρ / 2 * sum((V) .^ 2) / num_s + mean(objs)
+        )
+
+        return total_loss,
+        st_new,
+        (
+            total_loss = total_loss,
+            mean_violations = mean(V),
+            new_max_violation = maximum(V),
```
Comment (klamike) [inline src/L2OALM.jl]:
```suggestion
            max_violation = maximum(V),
```

why `new_`?

File: src/L2OALM.jl [hunk 11]
```diff
@@ -1,5 +1,426 @@
...
+This function computes the maximum violation, mean violations, mean objectives, and total loss
+from the batch states.
+"""
+function _reconcile_alm_primal_state(batch_states::Vector{NamedTuple})
+    max_violation = maximum([s.new_max_violation for s in batch_states])
+    mean_violations = mean([s.mean_violations for s in batch_states])
+    mean_objs = mean([s.mean_objs for s in batch_states])
+    mean_loss = mean([s.total_loss for s in batch_states])
+    return (;
+        new_max_violation = max_violation,
```
Comment (klamike) [inline src/L2OALM.jl]:
```suggestion
    max_violation = maximum([s.max_violation for s in batch_states])
    mean_violations = mean([s.mean_violations for s in batch_states])
    mean_objs = mean([s.mean_objs for s in batch_states])
    mean_loss = mean([s.total_loss for s in batch_states])
    return (;
        max_violation = max_violation,
```

https://github.com/LearningToOptimize/L2OALM.jl/pull/2/files#r2224260241 cont.

File: src/L2OALM.jl [hunk 12]
```diff
@@ -1,5 +1,426 @@
...
+function _update_ALM_ρ!(
+    hpm_primal::Dict{Symbol,Any},
+    hpm_dual::Dict{Symbol,Any},
+    current_state::NamedTuple,
+)
+    if current_state.new_max_violation > hpm_primal.τ * hpm_primal.max_violation
+        hpm_primal[:ρ] = min(hpm_primal[:ρmax], hpm_primal[:ρ] * hpm_primal[:α])
+        hpm_dual[:ρ] = hpm_primal[:ρ]  # Ensure dual model uses the same ρ
+    end
+    hpm_primal.max_violation = current_state.new_max_violation
```
Comment (klamike) [inline src/L2OALM.jl]:
```suggestion
    if current_state.max_violation > hpm_primal.τ * hpm_primal.max_violation
        hpm_primal[:ρ] = min(hpm_primal[:ρmax], hpm_primal[:ρ] * hpm_primal[:α])
        hpm_dual[:ρ] = hpm_primal[:ρ]  # Ensure dual model uses the same ρ
    end
    hpm_primal.max_violation = current_state.max_violation
```

https://github.com/LearningToOptimize/L2OALM.jl/pull/2/files#r2224260241 cont.

File: src/L2OALM.jl [hunk 13]
```diff
@@ -1,5 +1,426 @@
...
+    return TrainingStepLoop(
+        LagrangianPrimalLoss(bm),
+        [(iter, current_state, hpm) -> iter >= 100 ? true : false],
+        Dict{Symbol,Any}(
+            :ρ => 1.0,
+            :ρmax => 1e6,
+            :τ => 0.8,
+            :α => 10.0,
+            :max_violation => 0.0,
+        ),
```
Comment (klamike) [inline src/L2OALM.jl]: Does Lux let you make these structs instead?

File: src/L2OALM.jl [hunk 14]
```diff
@@ -1,5 +1,426 @@
...
+    )
+end
+
+"""
+    _reconcile_alm_dual_state(batch_states::Vector{NamedTuple})
+
+Default function that reconciles the state of the dual model after processing a batch of data.
+This function computes the mean dual loss from the batch states.
+"""
+function _reconcile_alm_dual_state(batch_states::Vector{NamedTuple})
```
Comment (klamike) [inline src/L2OALM.jl]:
```suggestion
function _reconcile_dual_state(batch_states::Vector{NamedTuple})
```

`alm` can be removed since that is this whole repo :smile: (needs updates everywhere else, and for the primal version, update_rho, etc. too. let me know if you agree and I can add that commit)

File: src/L2OALM.jl [hunk 15]
```diff
@@ -1,5 +1,426 @@
...
+function _default_dual_loop(num_equal::Int)
+    return TrainingStepLoop(
+        LagrangianDualLoss(num_equal),
+        [(iter, current_state, hpm) -> iter >= 100 ? true : false],
+        Dict{Symbol,Any}(:max_dual => 1e6, :ρ => 1.0),
+        [],
+        _reconcile_alm_dual_state,
+        _pre_hook_dual,
+    )
+end
```
Comment (klamike) [inline src/L2OALM.jl]: How about exposing the hyperparameters as kwargs here? Same for the primal one.

File: src/L2OALM.jl [hunk 16]
```diff
@@ -1,5 +1,426 @@
...
+    num_batches = length(data)
+    current_state_primal = (;)
+    current_state_dual = (;)
+
+    # primal loop
+    while all(
+        stopping_criterion(
+            iter_primal,
+            current_state_primal,
+            training_step_loop_primal.hyperparameters,
```
Comment (klamike) [inline src/L2OALM.jl]: Is this some standard Lux API? our stopping criteria don't need the state nor hyperparameters

File: src/L2OALM.jl [hunk 17]
```diff
@@ -1,5 +1,426 @@
...
+    train_state_primal,
+    dual_model,
+    train_state_dual,
+    bm,
+)
+    # Forward pass for dual model
+    dual_hat_k, _ = dual_model(θ, train_state_dual.parameters, train_state_dual.states)
+
+    return (dual_hat_k,)
+end
```
Comment (klamike) [inline src/L2OALM.jl]: Is there some guidance for what to put in a "pre-hook" vs "loss" ? Do they get treated differently somehow?

**Review by klamike — CHANGES_REQUESTED** [review]
Awesome work! I have a few comments, nothing major

File: src/L2OALM.jl [hunk 18]
```diff
@@ -1,5 +1,426 @@
...
+- `reconcile_state`: Function to reconcile the state after processing a batch of data.
+"""
+mutable struct TrainingStepLoop
+    loss_fn::Function
+    stopping_criteria::Vector{Function}
+    hyperparameters::Dict{Symbol,Any}
+    parameter_update_fns::Vector{Function}
+    reconcile_state::Function
+    pre_hook::Function
+end
```
Comment (klamike) [inline src/L2OALM.jl]: Why `Vector {Function}` for `parameter_update_fns`, `stopping_criteria`? I think it only ever uses one.

I see for the dual case there is no `parameter_update_fn`. I guess`(x...) -> nothing` can work there..

File: test/runtests.jl [hunk 19]
```diff
@@ -1,6 +1,169 @@
...
+
+        nvar = model.meta.nvar
+        ncon = model.meta.ncon
+        nθ = length(model.θ)
+        num_equal = nbus * 2
+
+        Θ_train = randn(T, nθ, dataset_size) |> dev_gpu
+        Θ_test = randn(T, nθ, dataset_size) |> dev_gpu
+
+        primal_model = feed_forward_builder(nθ, nvar, [320, 320])
```
Comment (klamike) [inline test/runtests.jl:74]: Not sure where but we should eventually have some magic for this... something like `L2ONN.feed_forward(bm, input=:all_params, output=:all_vars, hidden_sizes=[320,320])`

File: test/runtests.jl [hunk 20]
```diff
@@ -1,6 +1,169 @@
...
+        backend = CPU(),
+        batch_size = 32,
+        dataset_size = 3200,
+        rng = Random.default_rng(),
+        T = Float64,
+    )
+        model, nbus, ngen, blines =
+            create_parametric_ac_power_model(filename; backend = backend, T = T)
+        bm_train = BNK.BatchModel(model, batch_size, config = BNK.BatchModelConfig(:full))
+        bm_test = BNK.BatchModel(model, dataset_size, config = BNK.BatchModelConfig(:full))
```
Comment (klamike) [inline test/runtests.jl:64]:
```suggestion
        bm_train = BNK.BatchModel(model, batch_size, config = BNK.BatchModelConfig(:viol_grad))
        bm_test = BNK.BatchModel(model, dataset_size, config = BNK.BatchModelConfig(:viol_grad))
```

`viol_grad` suffices, to avoid jprod and hessian storage

File: Project.toml [hunk 21]
```diff
@@ -3,11 +3,29 @@ uuid = "f31bfc7b-7b5d-4cc3-b76b-1af281ce159d"
...
+
+[sources]
+BatchNLPKernels = {url = "https://github.com/klamike/BatchNLPKernels.jl"}
```
Comment (klamike) [inline Project.toml]:
```suggestion
BatchNLPKernels = {url = "https://github.com/LearningToOptimize/BatchNLPKernels.jl"}
```

File: src/L2OALM.jl (hunk 3 above)
Comment (andrewrosemberg) [reply src/L2OALM.jl]:
```julia
mutable struct PrimalDualTrainer
    primal_model::Lux
    primal_training_state::
    dual_model::Lux
    dual_training_state::
    data::Dataloader

File: src/L2OALM.jl (hunk 17 above)
Comment (andrewrosemberg) [reply src/L2OALM.jl]: keep hooks but move primal and dual evaluation inside loop with `Chainerules.ignore_derivatives`

File: src/L2OALM.jl [hunk 22]
```diff
@@ -1,5 +1,373 @@
...
+
+        # Get previous dual predictions
+        dual_hat_k, _ = @ignore_derivatives dual_model(Θ, prev_dual_training_state.parameters, prev_dual_training_state.states)
+
+        # Separate bound and equality constraints
+        # # Forward pass for primal model
+        X̂, _ = trainer.primal_model(θ, primal_training_state.parameters, primal_training_state.states)
+        gh = constraints!(bm, X̂, Θ)
+
+        @ignore_derivatives gh = trainer.constraints(Θ, trainer.ps_constraints, trainer.st_constraints)
```
Comment (ivanightingale) [inline src/L2OALM.jl:159]: Are `constraints`, `ps_constraints`, `st_constraints` fields of `ALMTrainer`?
